### PR TITLE
feat(platform-server): add an API to transfer state from server

### PR DESCRIPTION
**Pull request by vikerman**
TransferState provides a shared store that is transferred from the
server to client. To use it import BrowserTransferStateModule from the
client app module and ServerTransferStateModule from the server app
module.

Design doc: https://docs.google.com/document/d/1Hv3VTRQC0H5yPe8ivJXaGst93yRAtD6nEFkQDLiK_IU/preview

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: packages/platform-browser/src/browser/transfer_store.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * ```
+ * const COUNTER_KEY: TransferKey<number> = {key: 'counter'};
+ * let value = 10;
+ *
+ * transferStore.set(COUNTER_KEY, value);
+ * ```
+ *
+ * @experimental
+ */
+export interface TransferKey<T> { readonly key: string; }
```
Comment (alxhub) [inline packages/platform-browser/src/browser/transfer_store.ts]: `StateKey`?

File: packages/platform-browser/src/browser/transfer_store.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * The values in the store are serialized/deserialized using JSON.stringify/JSON.parse.So only
+ * boolean, number, string, null and non-class objects will be serialized and deserialzied in a
+ * non-lossy manner.
+ *
+ * @experimental
+ */
+@Injectable()
+export class TransferStore {
+  private store: {[k: string]: {}} = {};
```
Comment (alxhub) [inline packages/platform-browser/src/browser/transfer_store.ts]: Nit: `{[k: string]: {}}` implies that there is a value for _every_ possible key.

`{[k: string]: {}|undefined}`

File: packages/platform-browser/src/browser/transfer_store.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  private onSerializeCallbacks: {[k: string]: () => {}} = {};
+
+  /** @internal */
+  static init(initState: {}) {
```
Comment (alxhub) [inline packages/platform-browser/src/browser/transfer_store.ts]: Okay to use constructors for injected classes.

File: packages/platform-browser/src/browser/transfer_store.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    const store = new TransferStore();
+    store.store = initState;
+    return store;
+  }
+
+  /**
+   * Get the value corresponding to a key. Return `defaultValue` if key is not found.
+   */
+  get<T>(key: TransferKey<T>, defaultValue: T): T {
+    return this.store[key.key] as T || defaultValue;
```
Comment (alxhub) [inline packages/platform-browser/src/browser/transfer_store.ts]: What happens if `T` is nullable?

File: packages/platform-browser/src/browser/transfer_store.ts (hunk 2 above)
Comment (alxhub) [inline packages/platform-browser/src/browser/transfer_store.ts]: Is the state guaranteed to be an object `{}`? Can't it also be a primitive value? Maybe this is a use case for `any`?

File: packages/platform-server/src/transfer_store.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {APP_ID, NgModule} from '@angular/core';
+import {DOCUMENT, TransferStore, ɵescapeHtml as escapeHtml} from '@angular/platform-browser';
+
+import {} from './server';
```
Comment (alxhub) [inline packages/platform-server/src/transfer_store.ts]: Whoops.

File: tools/public_api_guard/platform-server/index.d.ts [hunk 6]
```diff
@@ -41,5 +41,9 @@ export declare function renderModuleFactory<T>(moduleFactory: NgModuleFactory<T>
 export declare class ServerModule {
 }
 
+/** @experimental */
+export declare class ServerTransferStoreModule {
```
Comment (alxhub) [inline tools/public_api_guard/platform-server/index.d.ts]: `ServerTransferStateModule`?

**Review by alxhub — APPROVED** [review]

File: packages/platform-browser/src/browser/transfer_store.ts (hunk 2 above)
Comment (Toxicable) [reply packages/platform-browser/src/browser/transfer_store.ts]: `{}` can also be a primitive 
see https://goo.gl/2SWn85

File: packages/platform-browser/src/browser/transfer_store.ts (hunk 1 above)
Comment (vikerman) [reply packages/platform-browser/src/browser/transfer_store.ts]: Done

File: packages/platform-browser/src/browser/transfer_store.ts (hunk 2 above)
Comment (vikerman) [reply packages/platform-browser/src/browser/transfer_store.ts]: Done

Comment (vikerman) [reply packages/platform-browser/src/browser/transfer_store.ts]: {} encompasses primitive values also

File: packages/platform-browser/src/browser/transfer_store.ts (hunk 3 above)
Comment (vikerman) [reply packages/platform-browser/src/browser/transfer_store.ts]: Done

File: tools/public_api_guard/platform-server/index.d.ts (hunk 6 above)
Comment (vikerman) [reply tools/public_api_guard/platform-server/index.d.ts]: Renaming all Store to State

File: packages/platform-server/src/transfer_store.ts (hunk 5 above)
Comment (vikerman) [reply packages/platform-server/src/transfer_store.ts]: Done

File: packages/platform-browser/src/platform-browser.ts [hunk 7]
```diff
@@ -10,10 +10,12 @@ export {BrowserModule, platformBrowser} from './browser';
 export {Meta, MetaDefinition} from './browser/meta';
 export {Title} from './browser/title';
 export {disableDebugTools, enableDebugTools} from './browser/tools/tools';
+export {BrowserTransferStateModule, StateKey, TransferState} from './browser/transfer_state';
 export {By} from './dom/debug/by';
 export {DOCUMENT} from './dom/dom_tokens';
 export {EVENT_MANAGER_PLUGINS, EventManager} from './dom/events/event_manager';
 export {HAMMER_GESTURE_CONFIG, HammerGestureConfig} from './dom/events/hammer_gestures';
 export {DomSanitizer, SafeHtml, SafeResourceUrl, SafeScript, SafeStyle, SafeUrl, SafeValue} from './security/dom_sanitization_service';
+
```
Comment (mhevery) [inline packages/platform-browser/src/platform-browser.ts]: revert

File: packages/platform-browser/src/browser/transfer_state.ts [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ * ```
+ * const COUNTER_KEY: StateKey<number> = {key: 'counter'};
+ * let value = 10;
+ *
+ * transferState.set(COUNTER_KEY, value);
+ * ```
+ *
+ * @experimental
+ */
+export interface StateKey<T> { readonly key: string; }
```
Comment (mhevery) [inline packages/platform-browser/src/browser/transfer_state.ts]: brandeed `makeKey`

**Review by mhevery — APPROVED** [review]

**Comment by IgorMinar** [comment]
can you please rebase?

**Comment by mcferren** [comment]
can you please share a link to an example directory that uses this feature?

**Comment by feloy** [comment]
@mcferren I just published an article about the subject: https://medium.com/@feloy/using-transferstate-api-in-an-angular-5-universal-app-130f3ada9e5b

**Comment by mcferren** [comment]
Thanks @feloy 

but I am consistently getting the default null value when I console logging the this.transferState.get(BATCH_KEY, null);

my use case is not tied to resolving a route as you have implemented. You used onSerialize<T>(key: StateKey<T>, callback: () => T): void as a result of resolving a route?

is this.transferState.set(BATCH_KEY, batch); currently operational?

i have two functions in my injectable. One that awaits an ajax call then sets. And another function that gets and returns. Inside my Injectable file, above the @Injectable syntax, I define 
const BATCH_KEY = makeStateKey<any>('batch');

my component checks and calls the injectable function that sets if( !isPlatformBrowser( this.platformId ) ) { and it calls the injectable function that gets if( isPlatformBrowser( this.platformId ) ) { 

what am I doing wrong to consistently get the default null value?

**Comment by dottodot** [comment]
@feloy I've tried to follow your article but  `const found = this.transferState.hasKey(RESULT_KEY);` always returns false on the browser, making the transferState to not work. 

Also `this.transferState.get(RESULT_KEY, null);` returns `null`

I've even cloned your repo and that's the same.

Oddly though if I do `console.log(this.transferState);` at the beginning of my resolve it shows the data is in the store. It's as if `get` and `hasKey` don't work.

**Comment by feloy** [comment]
@dottodot @mcferren Have you tried with different browsers?

Unfortunately, on a Linux system, I can make it work with Firefox only.

**Comment by dottodot** [comment]
On mac just tried Chrome, Safari and Firefox and none of them work.

**Comment by dottodot** [comment]
I've figured it out. In my `main.ts` I had

```
platformBrowserDynamic()
  .bootstrapModule(AppBrowserModule)
  .catch(err => console.log(err));
```
but you should bootstrap the application only when the document is ready so like this.

```
document.addEventListener('DOMContentLoaded', () => {
  platformBrowserDynamic()
    .bootstrapModule(AppBrowserModule)
    .catch(err => console.log(err));
});
```
now the transferState works

**Comment by mcferren** [comment]
yes I can confirm this as well - thank you fellas

**Pull request closed without merging**